### PR TITLE
Avoid sending stop signal more than once

### LINKED ISSUE #275: gRPC networking error
The test `./tests/basic_grpc_test.sh` fails on master.

## Current Behavior

Error from the test: 
```
agourlay@agourlay-thinkpad:~/Workspace/qdrant$ ./tests/basic_grpc_test.sh
++ dirname ./tests/basic_grpc_test.sh
+ cd ./tests/../
+ QDRANT_HOST=localhost:6334
+ docker_grpcurl='docker run --rm -it --network=host -v /home/agourlay/Workspace/qdrant/src/tonic/proto:/proto fullstorydev/grpcurl -plaintext -import-path /proto -proto qdrant.proto'
+ docker run --rm -it --network=host -v /home/agourlay/Workspace/qdrant/src/tonic/proto:/proto fullstorydev/grpcurl -plaintext -import-path /proto -proto qdrant.proto -d '{
   "name": "test_collection"
}' localhost:6334 qdrant.Collections/Delete
ERROR:
  Code: Canceled
  Message: stream terminated by RST_STREAM with error code: CANCEL
```

Error in the service logs:

```
[2022-02-02T08:20:35Z DEBUG h2::proto::settings] received settings ACK; applying Settings { flags: (0x0), initial_window_size: 1048576, max_frame_size: 16384 } 
[2022-02-02T08:20:35Z DEBUG h2::codec::framed_read] received frame=Headers { stream_id: StreamId(1), flags: (0x4: END_HEADERS) } 
[2022-02-02T08:20:35Z DEBUG h2::codec::framed_read] received frame=Data { stream_id: StreamId(1), flags: (0x1: END_STREAM) } 
[2022-02-02T08:20:35Z DEBUG h2::codec::framed_write] send frame=WindowUpdate { stream_id: StreamId(0), size_increment: 983041 } 
thread 'tonic-6' panicked at 'called `Result::unwrap()` on an `Err` value: ServiceError { error: "Can't reach one of the workers: channel closed" }', lib/collection/src/collection.rs:392:31
stack backtrace:
   0:     0x560fe96369fc - std::backtrace_rs::backtrace::libunwind::trace::h09f7e4e089375279
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x560fe96369fc - std::backtrace_rs::backtrace::trace_unsynchronized::h1ec96f1c7087094e
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x560fe96369fc - std::sys_common::backtrace::_print_fmt::h317b71fc9a5cf964
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x560fe96369fc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::he3555b48e7dfe7f0
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x560fe965bb5c - core::fmt::write::h513b07ca38f4fb1b
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/fmt/mod.rs:1149:17
   5:     0x560fe962ef05 - std::io::Write::write_fmt::haf8c932b52111354
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/io/mod.rs:1697:15
   6:     0x560fe9638590 - std::sys_common::backtrace::_print::h195c38364780a303
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x560fe9638590 - std::sys_common::backtrace::print::hc09dfdea923b6730
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x560fe9638590 - std::panicking::default_hook::{{closure}}::hb2e38ec0d91046a3
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:211:50
   9:     0x560fe9638145 - std::panicking::default_hook::h60284635b0ad54a8
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:228:9
  10:     0x560fe9638c44 - std::panicking::rust_panic_with_hook::ha677a669fb275654
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:606:17
  11:     0x560fe9638720 - std::panicking::begin_panic_handler::{{closure}}::h976246fb95d93c31
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:502:13
  12:     0x560fe9636ea4 - std::sys_common::backtrace::__rust_end_short_backtrace::h38077ee5b7b9f99a
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:139:18
  13:     0x560fe9638689 - rust_begin_unwind
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:498:5
  14:     0x560fe7c90951 - core::panicking::panic_fmt::h35f3a62252ba0fd2
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panicking.rs:107:14
  15:     0x560fe7c90a43 - core::result::unwrap_failed::hb53671404b9e33c2
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/result.rs:1613:5
  16:     0x560fe8510392 - core::result::Result<T,E>::unwrap::hf1e8c4a817a0cd4d
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/result.rs:1295:23
  17:     0x560fe84f5471 - <collection::collection::Collection as core::ops::drop::Drop>::drop::h1e0e40d03f2a63fa
                               at /home/agourlay/Workspace/qdrant/lib/collection/src/collection.rs:392:9
  18:     0x560fe84f0197 - core::ptr::drop_in_place<collection::collection::Collection>::hbc3a944734700c1d
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ptr/mod.rs:188:1
  19:     0x560fe828a404 - alloc::sync::Arc<T>::drop_slow::he2bdb6dd4c75e128
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/alloc/src/sync.rs:1079:18
  20:     0x560fe828e513 - <alloc::sync::Arc<T> as core::ops::drop::Drop>::drop::h035cb1d3052d520d
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/alloc/src/sync.rs:1640:13
  21:     0x560fe828e49b - core::ptr::drop_in_place<alloc::sync::Arc<collection::collection::Collection>>::h646e7c6535016077
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ptr/mod.rs:188:1
  22:     0x560fe7f8fdfb - storage::content_manager::toc::TableOfContent::delete_collection::{{closure}}::h324a5078d7e70c2c
                               at /home/agourlay/Workspace/qdrant/lib/storage/src/content_manager/toc.rs:230:9
  23:     0x560fe7e6fad3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h2039fd5601e12627
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  24:     0x560fe7f90dfe - storage::content_manager::toc::TableOfContent::perform_collection_operation::{{closure}}::hd903e7f4a1541bf9
                               at /home/agourlay/Workspace/qdrant/lib/storage/src/content_manager/toc.rs:312:17
  25:     0x560fe7e70cc3 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h396cbc514aa7a717
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  26:     0x560fe7fa4e6d - <qdrant::tonic::api::collections_api::CollectionsService as qdrant::tonic::qdrant::collections_server::Collections>::delete::{{closure}}::h22880f09d0ac962b
                               at /home/agourlay/Workspace/qdrant/src/tonic/api/collections_api.rs:103:22
  27:     0x560fe7e74719 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h85bf5f24e2618886
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  28:     0x560fe7e5adeb - <core::pin::Pin<P> as core::future::future::Future>::poll::haec036c486bddbec
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/future.rs:119:9
  29:     0x560fe8027342 - <<qdrant::tonic::qdrant::collections_server::CollectionsServer<T> as tower_service::Service<http::request::Request<B>>>::call::DeleteSvc<T> as tonic::server::service::UnaryService<qdrant::tonic::qdrant::DeleteCollection>>::call::{{closure}}::haa237a8cf858b61a
                               at /home/agourlay/Workspace/qdrant/src/tonic/qdrant.rs:499:52
  30:     0x560fe7e768b9 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hb76962349e4d5ed3
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  31:     0x560fe7e5adeb - <core::pin::Pin<P> as core::future::future::Future>::poll::haec036c486bddbec
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/future.rs:119:9
  32:     0x560fe7fb2388 - tonic::server::grpc::Grpc<T>::unary::{{closure}}::haec57f18286a58f7
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.6.2/src/server/grpc.rs:198:24
  33:     0x560fe7e79867 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hf4b4ba3ef6d910b6
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  34:     0x560fe8025689 - <qdrant::tonic::qdrant::collections_server::CollectionsServer<T> as tower_service::Service<http::request::Request<B>>>::call::{{closure}}::hdb8bd3b519b95dc4
                               at /home/agourlay/Workspace/qdrant/src/tonic/qdrant.rs:514:35
  35:     0x560fe7e735a7 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h7046c65515d72337
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/mod.rs:80:19
  36:     0x560fe7e5a7ab - <core::pin::Pin<P> as core::future::future::Future>::poll::h103727fa7032d4c7
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/future.rs:119:9
  37:     0x560fe7e572da - <F as futures_core::future::TryFuture>::try_poll::h1e515c36a157e215
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.19/src/future.rs:82:9
  38:     0x560fe805ae66 - <futures_util::future::try_future::into_future::IntoFuture<Fut> as core::future::future::Future>::poll::h9cb87525e37d034a
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/try_future/into_future.rs:34:9
  39:     0x560fe7ced9d8 - <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll::h3cabf74e2ba935f9
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/future/map.rs:55:37
  40:     0x560fe7dc95c6 - <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll::h412da53d61242d57
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/lib.rs:91:13
  41:     0x560fe7f46566 - <futures_util::future::try_future::MapErr<Fut,F> as core::future::future::Future>::poll::ha1090a79d45763dc
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/lib.rs:91:13
  42:     0x560fe7cf54f2 - <futures_util::future::either::Either<A,B> as core::future::future::Future>::poll::h5bb324f3f2f5ec0b
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/either.rs:89:32
  43:     0x560fe7cf522a - <F as futures_core::future::TryFuture>::try_poll::h7d395f8cc6d33bee
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.19/src/future.rs:82:9
  44:     0x560fe805ae16 - <futures_util::future::try_future::into_future::IntoFuture<Fut> as core::future::future::Future>::poll::h8b9c90a9571cbf40
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/try_future/into_future.rs:34:9
  45:     0x560fe7cedd78 - <futures_util::future::future::map::Map<Fut,F> as core::future::future::Future>::poll::h421547af36ebb6dc
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/future/map.rs:55:37
  46:     0x560fe7dc9526 - <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll::h359e5a120bf52f47
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/lib.rs:91:13
  47:     0x560fe7f464c6 - <futures_util::future::try_future::MapErr<Fut,F> as core::future::future::Future>::poll::h4896632c0a7ac74d
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/lib.rs:91:13
  48:     0x560fe7cf547d - <futures_util::future::either::Either<A,B> as core::future::future::Future>::poll::h19a1d96344bdfa71
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.19/src/future/either.rs:90:33
  49:     0x560fe802c096 - <tonic::transport::service::grpc_timeout::ResponseFuture<F> as core::future::future::Future>::poll::h0a9f5e50d1230ec6
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.6.2/src/transport/service/grpc_timeout.rs:88:38
  50:     0x560fe7cf7955 - <tower::util::either::Either<A,B> as core::future::future::Future>::poll::he8dfef85e8678b53
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tower-0.4.8/src/util/either.rs:71:57
  51:     0x560fe7f96d94 - <tonic::transport::server::recover_error::ResponseFuture<F> as core::future::future::Future>::poll::h2a5f205b5c4e8ec2
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.6.2/src/transport/server/recover_error.rs:63:20
  52:     0x560fe7fc9120 - <tonic::transport::server::SvcFuture<F> as core::future::future::Future>::poll::hf0ac75c7706e5d2a
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tonic-0.6.2/src/transport/server/mod.rs:819:50
  53:     0x560fe7e5a8cb - <core::pin::Pin<P> as core::future::future::Future>::poll::h1e3e2768fb3d701e
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/future/future.rs:119:9
  54:     0x560fe7d58900 - hyper::proto::h2::server::H2Stream<F,B>::poll2::h36098a7697818305
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.16/src/proto/h2/server.rs:438:37
  55:     0x560fe7d5e413 - <hyper::proto::h2::server::H2Stream<F,B> as core::future::future::Future>::poll::h8ca6d1a25db10961
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.14.16/src/proto/h2/server.rs:527:9
  56:     0x560fe7ec4bdd - tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}::he453f95460012f25
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/core.rs:161:17
  57:     0x560fe7ea0445 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h39f1dc149b46be5f
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/loom/std/unsafe_cell.rs:14:9
  58:     0x560fe7ec3f27 - tokio::runtime::task::core::CoreStage<T>::poll::h0c20160435b358ee
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/core.rs:151:13
  59:     0x560fe7e939ca - tokio::runtime::task::harness::poll_future::{{closure}}::h9b36e2240b962f3e
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:461:19
  60:     0x560fe7df8433 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::he9070225b1c50b2c
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panic/unwind_safe.rs:271:9
  61:     0x560fe7d00620 - std::panicking::try::do_call::h1c292301249f4be3
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:406:40
  62:     0x560fe7d038bb - __rust_try
  63:     0x560fe7cfe90a - std::panicking::try::h6886480be24ce9cf
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:370:19
  64:     0x560fe7da862a - std::panic::catch_unwind::h7385a51e90e6bcf2
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panic.rs:133:14
  65:     0x560fe7e928eb - tokio::runtime::task::harness::poll_future::h1369a3509a1464c0
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:449:18
  66:     0x560fe7e93e8a - tokio::runtime::task::harness::Harness<T,S>::poll_inner::h200e399a3b2114db
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:98:27
  67:     0x560fe7e9b093 - tokio::runtime::task::harness::Harness<T,S>::poll::h35a98021fc781d95
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:53:15
  68:     0x560fe7cb7220 - tokio::runtime::task::raw::poll::h8a73e1b650f6948c
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/raw.rs:113:5
  69:     0x560fe952563f - tokio::runtime::task::raw::RawTask::poll::h764fce7f7a8a88ff
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/raw.rs:70:18
  70:     0x560fe94fd272 - tokio::runtime::task::LocalNotified<S>::run::h63139b17874ae262
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/mod.rs:343:9
  71:     0x560fe9531602 - tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}::hef5b52f1a491fcf5
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:425:13
  72:     0x560fe9507ebd - tokio::coop::with_budget::{{closure}}::hab2505f353fd1d03
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:102:9
  73:     0x560fe95463cb - std::thread::local::LocalKey<T>::try_with::h8999fb026c88733c
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/thread/local.rs:399:16
  74:     0x560fe954560e - std::thread::local::LocalKey<T>::with::hf2d950bf7d0f5bdb
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/thread/local.rs:375:9
  75:     0x560fe953150b - tokio::coop::with_budget::hb80eb588582ec605
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:95:5
  76:     0x560fe953150b - tokio::coop::budget::h86a40246b8e5d83f
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/coop.rs:72:5
  77:     0x560fe953150b - tokio::runtime::thread_pool::worker::Context::run_task::h6f4ee77dd4da4aa8
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:424:9
  78:     0x560fe9530de6 - tokio::runtime::thread_pool::worker::Context::run::h219cf8f0e7440075
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:391:24
  79:     0x560fe9530be3 - tokio::runtime::thread_pool::worker::run::{{closure}}::hadfc2be5f7a93986
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:376:17
  80:     0x560fe94d6817 - tokio::macros::scoped_tls::ScopedKey<T>::set::h818de873664f9b32
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/macros/scoped_tls.rs:61:9
  81:     0x560fe9530b0f - tokio::runtime::thread_pool::worker::run::hd04c76ea2e43e292
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:373:5
  82:     0x560fe953098b - tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}::hb844fca8cdd080d4
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/thread_pool/worker.rs:352:45
  83:     0x560fe952f96a - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h00a3c744c96710d8
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/blocking/task.rs:42:21
  84:     0x560fe9515e2e - tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}::h0ea21afd96170c7b
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/core.rs:161:17
  85:     0x560fe94ec905 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h30deaedda7cc2dc4
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/loom/std/unsafe_cell.rs:14:9
  86:     0x560fe9515cd8 - tokio::runtime::task::core::CoreStage<T>::poll::hb2c4ef63adf1e3fc
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/core.rs:151:13
  87:     0x560fe953a38e - tokio::runtime::task::harness::poll_future::{{closure}}::hf6896fdf38ad646b
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:461:19
  88:     0x560fe9500ad3 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::heeeb933890adbb18
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panic/unwind_safe.rs:271:9
  89:     0x560fe955b1c2 - std::panicking::try::do_call::h4bc4e4b0c4632a8f
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:406:40
  90:     0x560fe955bebb - __rust_try
  91:     0x560fe955a8ca - std::panicking::try::h733dfee88aa12514
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:370:19
  92:     0x560fe953a97b - std::panic::catch_unwind::h58b323c70cf494d3
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panic.rs:133:14
  93:     0x560fe953a20c - tokio::runtime::task::harness::poll_future::hfe1045d27e85c1ff
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:449:18
  94:     0x560fe95383ec - tokio::runtime::task::harness::Harness<T,S>::poll_inner::h16d47fdaa012707b
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:98:27
  95:     0x560fe9538254 - tokio::runtime::task::harness::Harness<T,S>::poll::hf59ba16689d3c1d5
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/harness.rs:53:15
  96:     0x560fe9525802 - tokio::runtime::task::raw::poll::h65ccdb1b8863463b
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/raw.rs:113:5
  97:     0x560fe952563f - tokio::runtime::task::raw::RawTask::poll::h764fce7f7a8a88ff
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/raw.rs:70:18
  98:     0x560fe94fd324 - tokio::runtime::task::UnownedTask<S>::run::ha864ebd1e946f123
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/task/mod.rs:380:9
  99:     0x560fe94dc4f7 - tokio::runtime::blocking::pool::Task::run::hbd87c02bd85361da
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/blocking/pool.rs:91:9
 100:     0x560fe94de995 - tokio::runtime::blocking::pool::Inner::run::hee7cfa537aac9b0a
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/blocking/pool.rs:314:17
 101:     0x560fe94dddbc - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::h52190e01a585a06c
                               at /home/agourlay/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.16.1/src/runtime/blocking/pool.rs:294:17
 102:     0x560fe94b2d4c - std::sys_common::backtrace::__rust_begin_short_backtrace::hdfdca60894b996ad
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys_common/backtrace.rs:123:18
 103:     0x560fe953b911 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::ha90f37b08d5bc6a1
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/thread/mod.rs:484:17
 104:     0x560fe9500b11 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf5243083eeffc66a
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panic/unwind_safe.rs:271:9
 105:     0x560fe955b16c - std::panicking::try::do_call::h402db4f5edaf357d
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:406:40
 106:     0x560fe955bebb - __rust_try
 107:     0x560fe955ad31 - std::panicking::try::ha46a83dd14e5e95b
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:370:19
 108:     0x560fe953aaf1 - std::panic::catch_unwind::ha7649e2bfdb77139
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panic.rs:133:14
 109:     0x560fe953b725 - std::thread::Builder::spawn_unchecked::{{closure}}::hca2d4914265c8c38
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/thread/mod.rs:483:30
 110:     0x560fe94e48cf - core::ops::function::FnOnce::call_once{{vtable.shim}}::h9c9be3dd3b8d668b
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/ops/function.rs:227:5
 111:     0x560fe963d7c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hcbc6d2d80772be64
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/alloc/src/boxed.rs:1694:9
 112:     0x560fe963d7c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h9bffa2ca65a1d6e6
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/alloc/src/boxed.rs:1694:9
 113:     0x560fe963d7c3 - std::sys::unix::thread::Thread::new::thread_start::ha678a8b0caec8f55
                               at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/sys/unix/thread.rs:106:17
 114:     0x7f542fa93927 - start_thread
                               at ./nptl/./nptl/pthread_create.c:435:8
 115:     0x7f542fb239e4 - __clone
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:100
 116:                0x0 - <unknown>

```

**Interestingly, it does not always fail, more like half of the time I would say.**

## Steps to Reproduce

on master

1. start server with `cargo run --bin qdrant --features grpc`
2. run test with `./tests/basic_grpc_test.sh`

## Expected Behavior

No error running the test.

## Possible Solution

It seems to be caused by https://github.com/qdrant/qdrant/pull/273 as I tested the code before the merge and it seems to work.

### PR DESCRIPTION
**Pull request by e-ivkov**
Closes #275 

### Bug 
The problem was in the fact that stop signal was sent twice on collection deletion, which led to the signal being sent to an already closed channel (as worker stops after the first message) on the second attempt.

### Why it broke after channels migration
- In async-channels you could clone the `Receiver`
- So the sending will fail only if all receivers were dropped
- `UpdateHandler` that spawns worker threads, cloned the receiver to put it into the worker thread, but never used its own receiver
- So when the `Collection` is dropped and worker threads are ordered to close - their communication channel will never close as there was always a receiver in `UpdateHandler`

Now:
- In tokio ordinary unbounded channel receivers cannot be cloned
- To clone a receiver you would need a broadcast channel
- So in the `UpdateHandler` I replaced receiver with `Option<UnboundedReceiver>` as I noticed that this part of the channel is never used after it is cloned into the worker
- This led to us discovering that we send the stop signal after the worker thread was dead

### Changes
1. Stop signal will be sent only on Collection drop.
2. Drop is explicitly set to happen before removing collection directory.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable? - **Not applicable**
* [X] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by generall — APPROVED** [review]

**Review by agourlay — APPROVED** [review]
LGTM :+1: 

Better to wait for Andrey's approval though.

**Pull request merged**
